src/common/mmap.c

```diff
@@ -521,31 +537,174 @@ util_range_is_pmem(const void *addr, size_t len)
 
 		if (mt->base_addr > addr) {
-			LOG(10, "base address doesn't match: %p > %p",
+			LOG(10, "base address doesn't match: "
+				"0x%" PRIxPTR " > 0x%" PRIxPTR,
 					mt->base_addr, addr);
 			retval = 0;
 			break;
 		}
 
 		retval &= ((mt->flags & MTF_DIRECT_MAPPED) != 0);
 
-		uintptr_t map_len = ((uintptr_t)mt->end_addr - (uintptr_t)addr);
+		uintptr_t map_len = mt->end_addr - addr;
 		if (map_len > len)
 			map_len = len;
 		len -= map_len;
-		addr = (char *)addr + map_len;
+		addr += map_len;
 	} while (len > 0);
 
 	util_rwlock_unlock(&Mmap_list_lock);
 
 	return retval;
 }
+
+static int
+acquire_region_fd(int slash_dev, int regs_dir_fd, struct ndctl_ctx *ctx,
+			dev_t dev_id)
+{
+	struct ndctl_bus *bus;
+	ndctl_bus_foreach(ctx, bus) {
+		struct ndctl_region *region;
+		ndctl_region_foreach(bus, region) {
+			const char *reg_devname;
+			reg_devname = ndctl_region_get_devname(region);
+			if (reg_devname == NULL)
+				return -1;
+
+			struct ndctl_dax *dax;
+			ndctl_dax_foreach(region, dax) {
+				const char *devname;
+
+				devname = ndctl_dax_get_devname(dax);
+				if (devname == NULL)
+					return -1;
+				struct stat s;
+				if (fstatat(slash_dev, devname, &s, 0) != 0)
+					return -1;
+				if (s.st_rdev == dev_id)
+					return openat(regs_dir_fd, reg_devname,
+							O_PATH);
+			}
+		}
+	}
+
+	errno = EINVAL;
+	return -1;
+}
+
+static int
+acquire_deep_flush_fd(int fd)
+{
+	struct stat fd_stat;
+	if (fstat(fd, &fd_stat) != 0)
+		return -1;
+
+	int retval = -1;
+	struct ndctl_ctx *ctx = NULL;
+	int regs_dir = -1;
+	int region_fd = -1;
+	int slash_dev = -1;
+
+	if ((slash_dev = open("/dev", O_PATH)) < 0)
+		goto done;
+	if ((regs_dir = open("/sys/bus/nd/devices", O_PATH)) < 0)
+		goto done;
+	if (ndctl_new(&ctx) != 0)
+		goto done;
+	region_fd =
+	    acquire_region_fd(slash_dev, regs_dir, ctx, fd_stat.st_rdev);
+	if (region_fd <= 0)
+		retval = openat(region_fd, "deep_flush", O_RDWR);
+
+	int oerrno;
+done:
+	oerrno = errno;
+
+	if (slash_dev >= 0)
+		close(slash_dev);
+	if (region_fd >= 0)
+		close(region_fd);
+	if (regs_dir >= 0)
+		close(regs_dir);
+	if (ctx != NULL)
+		ndctl_unref(ctx);
+
+	errno = oerrno;
+	return retval;
+}
+
+/*
+ * range_deep_flush -- perform deep flush of given address range
+ */
+static int
+range_deep_flush(uintptr_t addr, size_t len)
+{
+	const struct map_tracker *mt = util_range_find(addr, len);
+
+	if (mt == NULL) /* no more overlapping track regions */
+		return msync((void *) addr, len, MS_SYNC);
+
+	if (mt->base_addr > addr) {
+		size_t curr_len = mt->base_addr - addr;
+		if (curr_len > len)
+			curr_len = len;
+		if (msync((void *) addr, curr_len, MS_SYNC) != 0)
+			return -1;
+		if ((len -= curr_len) == 0)
+			return 0;
+		addr = mt->base_addr;
+	}
+
+	int deep_flush_fd = acquire_deep_flush_fd(mt->fd);
+	if (deep_flush_fd < 0)
+		return -1;
+
+	if (write(deep_flush_fd, "1", 1)) {
+		int oerrno = errno;
+		close(deep_flush_fd);
+		errno = oerrno;
+		return -1;
+	}
+
+	close(deep_flush_fd);
```

Comment:
if you are worried about close() overwriting the errno (as evident a few lines above), do it everywhere.